main.py

```diff
@@ -867,25 +900,46 @@ def greedyOptimization(layouts: tuple, scores: array, info: InfoWithTime = None)
     """Randomly switches letters in each of the layouts to see whether the layouts can be improved this way."""
 
     optimizedLayouts = dict(zip(layouts, scores))
-    asciiArray = getAsciiArray()
     bigrams = getBigrams(''.join(sorted(layouts[0])))
+    if USE_CFFI:
+        bigr_count = len(bigrams)
+        bigr = ffi.new("Bigram[]", bigr_count)
+        for i, b in enumerate(bigrams):
+            bigr[i].letter1AsciiCode = b.letter1AsciiCode
+            bigr[i].letter2AsciiCode = b.letter2AsciiCode
+            bigr[i].frequency = b.frequency
+    else:
+        asciiArray = getAsciiArray()
     if DEBUG_MODE:
         print(f'DEBUG: Greedy optimization with {len(layouts)} layouts')
     else:
         info.set_status(f'{len(layouts)} layouts] [Greedy optimization')
     for layout, score in zip(layouts, deepcopy(scores)):
         optimizing = True
-        while optimizing is True:
-            layoutPermutations = performLetterSwaps(layout)
-            for i, permutatedLayout in enumerate(layoutPermutations):
-                permutatedScore = testSingleLayout(
-                    permutatedLayout, asciiArray, bigrams)
-                if permutatedScore > score:
-                    layout = permutatedLayout
-                    score = permutatedScore
-                    break
-                elif i+1 == len(layoutPermutations):
-                    optimizing = False
+        if USE_CFFI:
+            while optimizing is True:
+                layoutPermutations = performLetterSwaps(layout)
+                for i, permutatedLayout in enumerate(layoutPermutations):
+                    char_list = permutatedLayout.encode('latin1')
+                    permutatedScore = test_single_layout(char_list, len(permutatedLayout), bigr, bigr_count, LINEAR_SCORE_LIST)
```

Comment:
Good point. I thought it might be inefficient to check the value of `USE_CFFI` over and over again within the `while` loop. But I didn't test it. So I guess this was a case of premature optimization.

Reply:
I have tested it. Checking the value of `USE_CFFI` inside the loop increases the run time from 5m56s to 6m13 (that's ≈ +5%). Is it worth changing it?

Reply:
Hmmmm that's unfortunate. I didn't think checking a boolean variable would add that much computation time.
Let's leave it the way you wrote it!